packages/editor/src/components/block-settings-menu/index.js

```diff
@@ -98,31 +122,32 @@ export class BlockSettingsMenu extends Component {
 								<BlockModeToggle
 									clientId={ firstBlockClientId }
 									onToggle={ onClose }
-									role="menuitem"
 								/>
 							) }
 							{ count === 1 && (
 								<BlockUnknownConvertButton
 									clientId={ firstBlockClientId }
-									role="menuitem"
 								/>
 							) }
 							{ count === 1 && (
 								<BlockHTMLConvertButton
 									clientId={ firstBlockClientId }
-									role="menuitem"
 								/>
 							) }
-							<BlockDuplicateButton
```

Comment:
One of the challenges was providing the `withDispatch` props to the KeyboardShortcuts component. Another is that KeyboardShortcuts has to render regardless of whether the settings menu is open or closed (so that the shortcut is always bound and available while in a block).  That meant KeyboardShortcuts couldn't be added to the MenuItems (which would have been ideal).

Because of those constraints KeyboardShortcuts exists in the BlockSettingsMenu, and to avoid duplication I moved the `withDispatch` functions there as well. After that there was very little left to the individual components, they were just functions that returned MenuItems (with a little bit of boolean logic).

@noisysocks did suggest the logic in the dispatch function could become an effect. I've since found another complication though, that BlockSettingsMenu is also rendered in BlockMultiControls and receives different props from that parent component, so that would also have to be considered.

I thought this seemed to be the best trade-off.

Reply:
Fair enough 👍